Shift3/boilerplate-client-react
# Users can now create agents (and interact with what's been cached) while offline

@DropsOfSerenity Okay. I have added everything that I felt was important. As we discussed, if you can fix the stuff related to the caching side of things, that would be great. I will take care of the pagination-related issues once you are done. If you have any questions, feel free to let me know via Slack.

## Changes
1. User can now use the app while offline
2. Pages that are visited while the user is online will be inserted into the dynamic cache (found in the Cache Storage, which can be found in Chrome's Developer Tools). When the user goes offline and tries to access that page, they will be able to access it. If a page wasn't cached, the `offline.html` page will be shown instead.
3. The agents that the user sees on on the AgentListView will be stored into indexed db (via the `idb` package).
4. The user will be able to go to the UpdateAgentView while offline if the associated agent has been stored into indexed db.
5. If the user creates a new agent while offline, that request will be stored by workbox's background-sync plugin and sent to the server once the user's internet connection is restored. Instead of inserting the associated agent while the user is still offline, it made more sense to just show an endless info notification that serves as a visual queue of all the new agents that the user created while offline.
6. Installed various workbox packages in order to set up the offline caching functionality
7. Changed the "start_url" in the manifest.json file from "." to "./index.html"
8. Added a `sw.ts` file. This is the file that instructs the service worker on how to handle different requests. It will be used to generate the `sw.js` file when the app is built.
9. Registered the generated service worker via App.tsx
10. Added a utility.ts file that handles all of the operations related to manipulating the Indexed DB
11. Added endless error and info notifications
12. Moved the `getRandomNumber` method out of the NetworkDetector component so that it could be used elsewhere
13. Added a `getConnectionStatus` method to the NetworkDetector component's file
14. Added a `workbox-config.js` file

## Purpose
An admin should be able to create Agents even without a connection.

## Approach
I made use of `Workbox` and its various plugins to add the associated offline functionality to the app.

## The Issue
Previously, the only files and data that were initially cached in the cache called 'dynamic' were the following:
- the static bundle.js file
- /event-token
- /notification/?read__isnull=true

Now, the static bundle is no where to be found and it looks like all of the files in the bundle are being cached separately. Here's a screenshot:

![too many items in the dynamic cache](https://user-images.githubusercontent.com/2876874/222008996-b7a661e1-4350-4851-8bb9-6df6cf525c4d.png)

Here's a couple of other issues:

1. The `offline.html` file is no longer being used when the user visits pages that haven't yet been cached yet. At the moment, it gives the standard "This site can't be reached" page.

2. Opening an agent while online no longer works

3. The AgentListView won't be populated with agents while offline when there are agents in the indexed db.  

## Things that aren't broken

1. The precache cache is consistent with what I remember. The app is precaching all of the files in the public directory except for the service worker files.

<img width="850" alt="precached-files" src="https://user-images.githubusercontent.com/2876874/222237661-d63c8895-3e4e-4d0e-a005-4a51b0c0ed33.png">

## Learning

This is the video that gave me a jump start in integrating the newest version of Workbox into our boilerplate:

[Workbox TypeScript - Unpacking the Workbox](https://www.youtube.com/watch?v=sOq92prx00w)

These resources were very helpful as well:

[workbox-background-sync](https://developer.chrome.com/docs/workbox/modules/workbox-background-sync/)

[Managing fallback responses](https://developer.chrome.com/docs/workbox/managing-fallback-responses/)

## Demo

https://user-images.githubusercontent.com/2876874/222246209-37361895-9ead-490c-9e63-88f0d99173e8.mov

## Testing Steps
1. Pull in the changes to your local copy of this branch and run it alongside the dj_starter_demo repo.
2. Run the app as usual. The service worker build step has been added to both the start and build scripts.
3. Start the app and observe Chrome's Developer Tools. Look at the following areas on the left: Service Workers, Indexed DB, and Cache Storage.
4. In the "Service Workers" tab, enable the "Update on reload" option. This will cause the newly built service worker to be used by the app when the app is refreshed (ex. hitting on the refresh button in the browser). Before refreshing, you should delete whatever is inside of the Cache Storage tab. You should delete the dynamic and workbox-recache-v2 items. You should also delete the agents that have been stored in indexed db. In the IndexedDB tab, there will be an agents-store and an agents item inside of it. Right click on the agents item and click "Clear". When you refresh, you should see that the service worker has been received at the time that you refreshed the app. Watch the demo to get a better idea of the workflow that I'm using.

![service worker area](https://user-images.githubusercontent.com/2876874/222240121-c604e5b3-773b-477e-9284-fde309fe1650.png)
